File: ios-native/NeedsVsWants/Data/CurrencyFormatter.swift
import Foundation

/// Formats currency-immune cents into a display string per the chosen currency.
///
/// Improvement over the old `ios/` port: uses `NumberFormatter(.currency)` with
/// the currency's locale, so symbol placement, grouping, and — crucially —
/// minor-unit rules are correct. The old port hand-rolled `\(symbol) \(whole).\(c2)`
/// which produced wrong output for JPY (yen has no minor units).
enum CurrencyFormatter {

    static func format(cents: Int64, currency: CurrencyOption) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.locale = currency.locale
        formatter.currencyCode = currency.code

        if currency == .jpy {
            // JPY (and other zero-minor currencies): whole units, no decimals.
            formatter.maximumFractionDigits = 0
            formatter.minimumFractionDigits = 0
            return formatter.string(from: NSNumber(value: cents / 100)) ?? ""
        }

        return formatter.string(from: NSNumber(value: Double(cents) / 100.0)) ?? ""
    }
}

File: expo/src/data/CurrencyFormatter.ts
/**
 * Formats currency-immune cents into a display string per the chosen currency.
 *
 * Faithful port of CurrencyFormatter.swift. Uses Intl.NumberFormat with the
 * currency's locale, so symbol placement, grouping, and — crucially —
 * minor-unit rules are correct. JPY (and other zero-minor currencies) show
 * whole units with no decimals.
 */
import { CURRENCIES, type CurrencyCode } from "./schema";

const formatterCache = new Map<string, Intl.NumberFormat>();

function getFormatter(code: CurrencyCode): Intl.NumberFormat {
  const cached = formatterCache.get(code);
  if (cached) return cached;

  const meta = CURRENCIES[code];
  const formatter = new Intl.NumberFormat(meta.locale, {
    style: "currency",
    currency: code,
    minimumFractionDigits: meta.zeroMinor ? 0 : 2,
    maximumFractionDigits: meta.zeroMinor ? 0 : 2,
  });
  formatterCache.set(code, formatter);
  return formatter;
}

export function formatCents(cents: number, currency: CurrencyCode): string {
  const meta = CURRENCIES[currency];
  // Zero-minor currencies: whole units, no decimals.
  const units = meta.zeroMinor ? Math.trunc(cents / 100) : cents / 100;
  return getFormatter(currency).format(units);
}
